Load-time block format validation

rvagg opened the pull request:
Includes #57, the first two commits. The changes so far are only in `LoadNode` and in the tests.

Adds in some simple helper functions into the tests to help manually build CBOR blobs that'll load and trigger various cases. Only got as far as ensuring that the bitmap and the number of elements aren't out of alignment. That can get stricter with #54 if we put in precisely the size of bitmap that we need for the array of elements.

Other things I'm working on in here are listed at the bottom of hamt_test.go. The aim is that this thing should refuse to load from blocks that smell funny (i.e. aren't exactly the right form). There should only be one way of representing this data and there should be no avenues for variation. That's an ideal and not strictly possible in the extreme sense for various reasons but we can get close.

rvagg commented:
Ready for review now. This includes 2 commits from #57, the `cleanChild()` changes. It adds a bunch of code to `LoadNode()` and pulls out the loading functionality into a new `loadNode()` so we can toggle on root/non-root for different validation. There's a minor change that gets into the new `cleanChild()` because we no longer need to do one bit of validation there now. There's a new method in uhamt.go for counting the total number in the bitfield. There's a big-ol' test case named `TestMalformedHamt` which pushes this around with different incoming CBOR blocks. Some utilities in hamt_util_test.go plus some context-specific ones inside the test let us manually build CBOR that forces validation errors in various ways. There's also sanity-checking along the way to make sure we're not just hitting "this is all just bad CBOR" type errors.

The HAMT has (almost) zero chill now about malformed blocks. If a block smells bad then it's rejected, no pretending. One item that could be checked but is not because it's so complicated and costly (for now) is to make sure that entries are in their right positions in the graph given their keys. If that's important it could be added later and called on demand. For now this is block-local only.

Stebalien commented on hamt.go:
```diff
@@ -23,7 +23,7 @@ const defaultBitWidth = 8
 // ErrNotFound is returned when a Find operation fails to locate the specified
 // key in the HAMT
-var ErrNotFound = fmt.Errorf("not found")
+var ErrNotFound = fmt.Errorf("Not found")
```
Errors should be lower case.

rvagg replied:
👍 thanks, fixed

Stebalien commented on hamt.go:
```diff
@@ -32,7 +32,13 @@ var ErrNotFound = fmt.Errorf("not found")
 // represented with the hash function used. Hash functions with larger byte
 // output increase the maximum theoretical depth of a HAMT.
-var ErrMaxDepth = fmt.Errorf("attempted to traverse hamt beyond max depth")
+var ErrMaxDepth = fmt.Errorf("Attempted to traverse HAMT beyond max-depth")
```
ditto errors being lower case (except for acronyms)

Stebalien commented on hamt.go:
```diff
@@ -302,15 +308,10 @@ func (p *Pointer) loadChild(ctx context.Context, ns cbor.IpldStore, bitWidth int
 	}
 
-	out, err := LoadNode(ctx, ns, p.Link)
+	out, err := loadNode(ctx, ns, p.Link, false, UseTreeBitWidth(bitWidth), UseHashFunction(hash))
```
We should ideally avoid using these functions for every load as this will lead to additional allocations. In the grand scheme of things, it might not be that important, but it would be better to just explicitly list out all relevant options.

Alternatively, we could create shared "config" object that we pass from node to node on construction.

rvagg replied:
I've added 2 more arguments to the function so they get passed in explicitly, it gets a bit verbose but it's not terrible as it is now. We're probably going to want `bucketSize` at some point too so this might start to get out of hand soon and pulling up to a config object might be a good refactor.

warpfork replied:
+1 to being somewhat perturbed about how often we're passing these parameters around.  (It's given me an itchy feeling in some other patches, too.)

But agree with this already being marked resolved; revisiting that and trying to make it more parsimonious would be another whole PR.

rvagg replied:
mm, I wanted to avoid public API breakages in doing all of this so this seems like something to save til later. tbh this function pattern for setting options is new to me and I have no idea how common it is in go-land. I can see why one might choose it as a pattern but equally it seems both safer and just easier to pass a struct with fields you care about instead `Options{BitWidth: 3, HashFunction: identity}`, which would also work nicely for these internal usages.

So, another PR for someone who wants to mess with the public API.

rvagg commented on hamt.go:
```diff
@@ -328,33 +329,79 @@ func (p *Pointer) loadChild(ctx context.Context, ns cbor.IpldStore, bitWidth int
+		if !((isLink && !isBucket) || (!isLink && isBucket)) {
+			return nil, ErrMalformedHamt
+		}
+		if isLink && ch.Link.Type() != cid.DagCBOR { // not dag-cbor
```
Possibly the most controversial item in here that I want to highlight. We're validating that all links to child HAMT nodes (not links it might contain as values, only HAMT nodes) are of codec DAG-CBOR. Just another check to get some assurance we're not being fed bogus data in the structure, it doesn't go very far but we're asserting that this HAMT has homogeneity in its block generation, for now. This could be changed in future if we allowed some variability in codec or wanted to have a migration path.

warpfork replied:
We could... make this another parameter?  I'd agree, anyway, that having the interior tree all be one codec is probably pretty reasonable.

The repo name doesn't have "cbor" in it; that's my only reason for pause here in making it hardcoded to that value.

(Otoh, there's... several other details and import statements that seem to imply this system is already cbor-only, and the UnmarshalCBOR methods, and... hrm.  Maybe the repo name should just *also* indicate cbor-specificness, if it's already quietly the case.)

Stebalien replied:
This seems like a reasonable check for now, but I agree we should, ideally, make it more generic.

I mean, ideally we'd _save_ the codec/hash-function, etc. when loading a node, then restore it when encoding. Then, we'd just check "do all linked children match their parents?". But at the moment, you're right. Everything is very DagCBOR specific.

rvagg commented:
Fixed CI bug with Go 1.11 and squashed this into just the two commits on the HEAD of this branch. Review should focus on those two, 04503fda1 particularly (the other just adds a `coverage` target to Makefile).

Stebalien commented on hamt.go:
```diff
@@ -328,33 +329,79 @@ func (p *Pointer) loadChild(ctx context.Context, ns cbor.IpldStore, bitWidth int
+				return nil, ErrMalformedHamt
+			}
+			for i := 1; i < len(ch.KVs); i++ {
+				if bytes.Compare(ch.KVs[i-1].Key, ch.KVs[i].Key) >= 0 {
```
Followup: we should probably be checking the hash prefix.
